### PR TITLE
Support new IDE protocol in app model

### LINKED ISSUE #2928: Throw instead of calling `Environment.Exit` in `DcpDependencyCheck`
If the cancellation token passed to `DcpDependencyCheck.EnsureDcpDependenciesAsync(...)` is canceled before completion, this will currently crash the process. Instead, it should clean up and propagate the cancellation up the stack.

This is causing testing to crash when the test builder is not used as intended. We should not terminate the process this way, or we should provide a way to opt-out of crashing the process. There are 4 `Environment.Exit` calls in `DcpDependencyCheck`. They should all be converted into exceptions which propagate up the stack. This would improve the debugging experience.

Are the error codes passed to `Environment.Exit` important?

### PR DESCRIPTION
**Pull request by karolz-ms**
Because it was decided that we should support the old version of the protocol for one preview release, the code inside `ApplicationExecutor` must do different things depending on the version of DCP it is talking to. As a result, I refactored a bit the DCP version/dependency check. Adding @ReubenBond who I was told is doing some work in this area right now.

Fixes #2928

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/3026)

### CODE REVIEW

File: src/Aspire.Hosting/Dcp/ApplicationExecutor.cs [hunk 1]
```diff
@@ -928,9 +938,10 @@ private void PreparePlainExecutables()
         }
     }
 
-    private void PrepareProjectExecutables()
+    private async Task PrepareProjectExecutablesAsync(CancellationToken cancellationToken)
     {
         var modelProjectResources = _model.GetProjectResources();
+        var dcpInfo = await dcpDependencyCheckService.GetDcpInfoAsync(cancellationToken).ConfigureAwait(false);
```
Comment (davidfowl) [inline src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:944]: Perfect! I need this information to get the container runtime 😄 for the host name in this file.

File: src/Aspire.Hosting/Dcp/DcpDependencyCheck.cs [hunk 2]
```diff
@@ -29,79 +31,96 @@ internal sealed partial class DcpDependencyCheck : IDcpDependencyCheckService
         _dcpOptions = dcpOptions.Value;
     }
 
-    public async Task EnsureDcpDependenciesAsync(CancellationToken cancellationToken = default)
+    public async Task<DcpInfo?> GetDcpInfoAsync(CancellationToken cancellationToken = default)
```
Comment (davidfowl) [inline src/Aspire.Hosting/Dcp/DcpDependencyCheck.cs:34]: Any reason to not make this cache the result?

File: src/Aspire.Hosting/Dcp/DcpDependencyCheck.cs [hunk 3]
```diff
@@ -29,79 +31,96 @@ internal sealed partial class DcpDependencyCheck : IDcpDependencyCheckService
...
+                return dcpInfo;
+            }
+            catch (Exception ex) when (ex is not DistributedApplicationException)
             {
-                return; // Best effort
+                Console.Error.WriteLine(string.Format(
+                    CultureInfo.InvariantCulture,
+                    Resources.DcpDependencyCheckFailedMessage,
+                    ex.ToString()
+                ));
```
Comment (mitchdenny) [inline src/Aspire.Hosting/Dcp/DcpDependencyCheck.cs:109]: Use logger instead.

File: src/Aspire.Hosting/Dcp/DcpDependencyCheck.cs [hunk 4]
```diff
@@ -29,79 +31,96 @@ internal sealed partial class DcpDependencyCheck : IDcpDependencyCheckService
...
+                Environment.Exit((int)DcpVersionCheckFailures.DcpVersionFailed);
```
Comment (mitchdenny) [inline src/Aspire.Hosting/Dcp/DcpDependencyCheck.cs:110]: Can we throw here instead of doing an exit.

**Comment by davidfowl** [comment]
@karolz-ms I crapped on your PR an integrated @ReubenBond 's changes https://github.com/dotnet/aspire/pull/2931.

**Comment by davidfowl** [comment]
@danegsta we are throwing here. We can tweak this in a follow up PR to set a specific exit code for unhandled exceptions (that was a good idea 😄).

**Review by davidfowl — APPROVED** [review]

**Pull request merged**